evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java

```diff
@@ -512,12 +533,167 @@ <T> T doGet(EVCacheKey evcKey , Transcoder<T> tc) throws EVCacheException {
             if (!throwExc) return null;
             throw new EVCacheException("Exception getting data for APP " + _appName + ", key = " + evcKey, ex);
         } finally {
-            final long duration = EVCacheMetricsFactory.getInstance().getRegistry().clock().wallTime()- start;
+            final long duration = EVCacheMetricsFactory.getInstance().getRegistry().clock().wallTime() - start;
             getTimer(Call.GET.name(), EVCacheMetricsFactory.READ, cacheOperation, status, tries, maxReadDuration.get().intValue(), client.getServerGroup()).record(duration, TimeUnit.MILLISECONDS);
-            if (log.isDebugEnabled() && shouldLog()) log.debug("GET : APP " + _appName + ", Took " + duration + " milliSec.");
+            if (log.isDebugEnabled() && shouldLog())
+                log.debug("GET : APP " + _appName + ", Took " + duration + " milliSec.");
+        }
+    }
+
+    public <T> CompletableFuture<T> get(String key, Transcoder<T> tc, ExecutorService executorService) throws EVCacheException {
+        if (null == key) throw new IllegalArgumentException("Key cannot be null");
+        final EVCacheKey evcKey = getEVCacheKey(key);
+        return doAsyncGet(evcKey, tc, executorService);
+    }
+
+    <T> CompletableFuture<T> doAsyncGet(EVCacheKey evcKey, Transcoder<T> tc, ExecutorService executorService) throws EVCacheException {
+        final boolean throwExc = doThrowException();
+        EVCacheClient client = buildEvCacheClient(throwExc);
+        if (client == null) {
+            return null;
+        }
+        final EVCacheEvent event = buildAndStartEvent(client, evcKey, throwExc);
+        final long start = EVCacheMetricsFactory.getInstance().getRegistry().clock().wallTime();
+        StringBuilder status = new StringBuilder(EVCacheMetricsFactory.SUCCESS);
+        StringBuilder cacheOperation = new StringBuilder(EVCacheMetricsFactory.YES);
+        AtomicInteger tries = new AtomicInteger(1);
+        final boolean hasZF = hasZoneFallback();
+        boolean throwEx = !hasZF && throwExc;
+        return getAsyncData(client, evcKey, tc, throwEx, hasZF, executorService)
+                .thenCompose(data -> handleRetry(data, evcKey, tc, client, hasZF, throwExc, event, executorService))
+                .thenApply(data -> handleData(data, event, evcKey, client, cacheOperation))
+                .exceptionally(ex -> handleException(ex, throwExc, event, evcKey))
+                .handle((data, ex) -> handleFinally(data, status, tries, client, cacheOperation, start));
+    }
+
+    private EVCacheClient buildEvCacheClient(boolean throwExc) throws EVCacheException {
+        EVCacheClient client = _pool.getEVCacheClientForRead();
+        if (client == null) {
+            incrementFastFail(EVCacheMetricsFactory.NULL_CLIENT, Call.COMPLETABLE_FUTURE_GET);
```

Comment:
method "buildEvCacheClient" seems generic (beyond just Completable-Future-Get). Semantically, it seems odd this method increments null client for CompletableFutureGet.
My recommendation - either take the Call enum as argument in the method, or not create this method and build new evcache inline within "doAsyncGet" method